secboot/tpm.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	t := &tpmSupport{tconn: tconn, lockoutAuth: lockoutAuth}
+
+	return t, nil
+}
+
+// StoreLockoutAuth saves the lockout authorization data in a file at the
+// path specified by filename.
+func (t *tpmSupport) StoreLockoutAuth(filename string) error {
+	if err := osutil.AtomicWriteFile(filename, t.lockoutAuth, 0600, 0); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetShimFiles verifies and sets the list of shim binaries.
+func (t *tpmSupport) SetShimFiles(pathList ...string) error {
+	if err := checkFilesPresence(pathList); err != nil {
+		return err
+	}
+	t.shimFiles = pathList
+	return nil
+}
+
+// SetBootloaderFiles verifies and sets the list of bootloader binaries.
+func (t *tpmSupport) SetBootloaderFiles(pathList ...string) error {
+	if err := checkFilesPresence(pathList); err != nil {
+		return err
+	}
+	t.bootloaderFiles = pathList
+	return nil
+}
+
+// SetKernelFiles verifies and sets the list of kernel binaries.
+func (t *tpmSupport) SetKernelFiles(pathList ...string) error {
+	if err := checkFilesPresence(pathList); err != nil {
+		return err
+	}
+	t.kernelFiles = pathList
+	return nil
+}
+
+func checkFilesPresence(pathList []string) error {
+	for _, p := range pathList {
+		if !osutil.FileExists(p) {
+			return fmt.Errorf("file %s does not exist", p)
+		}
+	}
+	return nil
+}
+
+var (
+	sbProvisionTPM = sb.ProvisionTPM
+)
+
+// Provision provisions the TPM. This will fail on a device that has
+// previously been provisioned.
+func (t *tpmSupport) Provision() error {
+	if err := sbProvisionTPM(t.tconn, sb.ProvisionModeFull, t.lockoutAuth); err != nil {
+		return fmt.Errorf("cannot provision TPM: %v", err)
+	}
+
+	return nil
+}
+
+// Close closes the TPM connection.
+func (t *tpmSupport) Close() error {
+	if err := t.tconn.Close(); err != nil {
+		return fmt.Errorf("cannot close TPM connection: %v", err)
+	}
+	return nil
+}
+
+func (t *tpmSupport) SetKernelCmdlines(cmdlines []string) {
+	t.kernelCmdlines = cmdlines
+}
+
+var (
+	sbSealKeyToTPM                  = sb.SealKeyToTPM
+	sbAddEFISecureBootPolicyProfile = sb.AddEFISecureBootPolicyProfile
+	sbAddSystemdEFIStubProfile      = sb.AddSystemdEFIStubProfile
+)
+
+// Seal seals the given key to the TPM and writes the sealed object to a file at the
+// path specified by keyPath. Additional data required for updating the authorization
+// policy is written to a file at the path specified by policyUpdatePath. This file
+// must live inside an encrypted volume protected by this key.
+func (t *tpmSupport) Seal(key []byte, keyPath, policyUpdatePath string) error {
+	pcrProfile := sb.NewPCRProtectionProfile()
+
+	// Add EFI secure boot policy profile
+	policyParams := sb.EFISecureBootPolicyProfileParams{
+		PCRAlgorithm:  tpm2.HashAlgorithmSHA256,
+		LoadSequences: make([]*sb.EFIImageLoadEvent, 0, len(t.shimFiles)),
+		// TODO:UC20: set SignatureDbUpdateKeystore to support key rotation
+	}
+
+	for _, shim := range t.shimFiles {
```

Comment:
Wonder whether we should have a hold of those paths inside the `tpmSupport` or rather have the right sequence be passed as a parameter?
```
type LoadChain struct {
    Shim string
    Bootloader string
    Kernel string
    CommandLine string
}
func (t *tpmSupport) Seal(chains []LoadChain) error {
...
}
```
Even if, it's probably a follow-up material.